thirdparty/DiskANN/include/aio_context_pool.h

```diff
@@ -0,0 +1,52 @@
+#pragma once
+
+#ifndef _WINDOWS
+
+#include <mutex>
+#include <queue>
+#include <libaio.h>
+#include <condition_variable>
+#include "utils.h"
+#include "concurrent_queue.h"
+
+class AioContextPool {
+ public:
+  explicit AioContextPool(size_t num_ctx, int max_events);
+
+  AioContextPool(const AioContextPool&) = delete;
+
+  AioContextPool& operator=(const AioContextPool&) = delete;
+
+  AioContextPool(AioContextPool&&) noexcept = delete;
+
+  AioContextPool& operator==(AioContextPool&&) noexcept = delete;
+
+  size_t max_events_per_ctx() {
+    return max_events_;
+  }
+
+  void push(io_context_t ctx) {
+    ctx_q_.push(ctx);
+    ctx_q_.push_notify_all();
+  }
+
+  io_context_t pop() {
+    io_context_t ctx = ctx_q_.pop();
+    while (!ctx) {
```

Comment:
Did get the idea. can you add some comments?
I guess this queue will not `blockingwait` but return 0?
can we use (!io_context_t) here?

Reply:
Also thinking, if a thread is hanging here waiting for the aio. Then the use try to stop the process, this thread will stuck here forever, so we cannot gracefully shutdown.